app/src/main/java/org/onionshare/android/ShareManager.kt

```diff
@@ -0,0 +1,168 @@
+package org.onionshare.android
+
+import android.app.Application
+import android.net.Uri
+import android.text.format.Formatter
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.onionshare.android.files.FileManager
+import org.onionshare.android.server.PORT
+import org.onionshare.android.server.SendFile
+import org.onionshare.android.server.SendPage
+import org.onionshare.android.server.WebserverManager
+import org.onionshare.android.tor.TorManager
+import org.onionshare.android.ui.ShareUiState
+import org.slf4j.LoggerFactory.getLogger
+import java.io.IOException
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private val LOG = getLogger(ShareManager::class.java)
+
+@Singleton
+class ShareManager @Inject constructor(
+    private val app: Application,
+    private val torManager: TorManager,
+    private val webserverManager: WebserverManager,
+    private val fileManager: FileManager,
+) {
+
+    private val _shareState = MutableStateFlow<ShareUiState>(ShareUiState.NoFiles)
+    val shareState: StateFlow<ShareUiState> = _shareState
+
+    @Volatile
+    private var startSharingJob: Job? = null
+
+    suspend fun addFiles(uris: List<Uri>) = withContext(Dispatchers.IO) {
+        if (uris.isEmpty()) return@withContext // user backed out of select activity
+
+        // not supporting selecting entire folders with sub-folders
+        val filesAdded = fileManager.addFiles(uris, shareState.value.files)
+        val totalSize = filesAdded.files.sumOf { it.size }
+        _shareState.value = ShareUiState.FilesAdded(filesAdded.files, totalSize)
+    }
+
+    fun removeFile(file: SendFile) {
+        val newList = shareState.value.files.filterNot { it == file }
+        if (newList.isEmpty()) {
+            _shareState.value = ShareUiState.NoFiles
+        } else {
+            val totalSize = newList.sumOf { it.size }
+            _shareState.value = ShareUiState.FilesAdded(newList, totalSize)
+        }
+    }
+
+    fun removeAll() {
+        _shareState.value = ShareUiState.NoFiles
+    }
+
+    suspend fun onStateChangeRequested() = when (shareState.value) {
+        is ShareUiState.FilesAdded -> startSharing()
+        is ShareUiState.Starting -> stopSharing()
+        is ShareUiState.Sharing -> stopSharing()
+        is ShareUiState.Complete -> startSharing()
+        is ShareUiState.Error -> startSharing()
+        ShareUiState.NoFiles -> error("Sheet button should not be visible with no files")
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
```

Comment:
This name isn't super encouraging. :smile: What are we doing that's delicate?

Reply:
It is because we are using the global scope here: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-global-scope/